route/netlink_path.go

```diff
@@ -0,0 +1,171 @@
+package route
+
+import (
+	"fmt"
+
+	bnet "github.com/bio-routing/bio-rd/net"
+	log "github.com/sirupsen/logrus"
+	"github.com/vishvananda/netlink"
+)
+
+// ProtoBio is the protocol number constant for bio from /etc/iproute2/rt_protos
+const ProtoBio = 45
+
+// NetlinkPath represents a path learned via Netlink of a route
+type NetlinkPath struct {
+	Dst      bnet.Prefix
+	Src      bnet.IP
+	NextHop  bnet.IP // GW
+	Priority int
+	Protocol int
+	Type     int
+	Table    int
+	Kernel   bool // True if the route is already installed in the kernel
+}
+
+// NewNlPathFromBgpPath creates a new NetlinkPath object from a BGPPath object
+func NewNlPathFromBgpPath(p *BGPPath) *NetlinkPath {
+	return &NetlinkPath{
+		Src:      p.Source,
+		NextHop:  p.NextHop,
+		Protocol: ProtoBio,
+		Kernel:   false,
+	}
+}
+
+// NewNlPathFromRoute creates a new NetlinkPath object from a netlink.Route object
+func NewNlPathFromRoute(r *netlink.Route, kernel bool) (*NetlinkPath, error) {
+	var src bnet.IP
+	var dst bnet.Prefix
+
+	if r.Src == nil && r.Dst == nil {
+		return nil, fmt.Errorf("Cannot create NlPath, since source and destination are both nil")
+	}
+
+	if r.Src == nil && r.Dst != nil {
+		dst = bnet.NewPfxFromIPNet(r.Dst)
+		if dst.Addr().IsIPv4() {
+			src = bnet.IPv4FromOctets(0, 0, 0, 0)
+		} else {
+			src = bnet.IPv6FromBlocks(0, 0, 0, 0, 0, 0, 0, 0)
+		}
+	}
+
+	if r.Src != nil && r.Dst == nil {
+		src, _ = bnet.IPFromBytes(r.Src)
+		if src.IsIPv4() {
+			dst = bnet.NewPfx(bnet.IPv4FromOctets(0, 0, 0, 0), 0)
+		} else {
+			dst = bnet.NewPfx(bnet.IPv6FromBlocks(0, 0, 0, 0, 0, 0, 0, 0), 0)
+		}
+	}
+
+	if r.Src != nil && r.Dst != nil {
+		src, _ = bnet.IPFromBytes(r.Src)
+		dst = bnet.NewPfxFromIPNet(r.Dst)
+	}
+
+	log.Warnf("IPFromBytes: %v goes to %v", r.Src, src)
+	log.Warnf("IPFromBytes: %v goes to %v", r.Dst, dst)
+
+	nextHop, _ := bnet.IPFromBytes(r.Gw)
+
+	return &NetlinkPath{
+		Dst:      dst,
+		Src:      src,
+		NextHop:  nextHop,
+		Priority: r.Priority,
+		Protocol: r.Protocol,
+		Type:     r.Type,
+		Table:    r.Table,
+		Kernel:   kernel,
+	}, nil
+}
+
+// Select compares s with t and returns negative if s < t, 0 if paths are equal, positive if s > t
+func (s *NetlinkPath) Select(t *NetlinkPath) int8 {
+	if !s.Dst.Equal(t.Dst) {
+		return 1
+	}
+
+	if s.NextHop.Compare(t.NextHop) > 0 {
+		return -1
+	}
+
+	if s.NextHop.Compare(t.NextHop) < 0 {
+		return 1
+	}
+
+	if s.Src.Compare(t.Src) > 0 {
+		return -1
+	}
+
+	if s.Src.Compare(t.Src) < 0 {
+		return 1
+	}
+
+	if s.Priority < t.Priority {
+		return -1
+	}
+
+	if s.Priority > t.Priority {
+		return 1
+	}
+
+	if s.Protocol < t.Protocol {
+		return -1
+	}
+
+	if s.Protocol > t.Protocol {
+		return 1
+	}
+
+	if s.Table < t.Table {
+		return -1
+	}
+
+	if s.Table > t.Table {
+		return 1
+	}
+
+	return 0
+}
+
+// ECMP determines if path s and t are equal in terms of ECMP
+func (s *NetlinkPath) ECMP(t *NetlinkPath) bool {
+	return true
+}
```

Comment:
#yolo?

Reply:
correct. ECMP is part of another PR. I will put that one here to `false` until it's implemented

Reply:
would like to click "Resolve conversation" if you think it's better now